RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+} 
+
+
+
+
+ 
+
+std::vector<HaloClusterCandidateHB> CSCHaloAlgo::GetHaloClusterCandidateHB(edm::Handle<EcalRecHitCollection>& ecalrechitcoll, edm::Handle<HBHERecHitCollection>& hbherechitcoll,float et_thresh_seedrh){
+
+  std::vector<HaloClusterCandidateHB> TheHaloClusterCandsHB;
+
+  reco::Vertex::Point vtx(0,0,0);
+
+  for(size_t ihit = 0; ihit<hbherechitcoll->size(); ++ ihit){
+    HaloClusterCandidateHB  clustercand;
+    
+    const HBHERecHit & rechit = (*hbherechitcoll)[ ihit ];
+    math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+    //Et condition
+    double rhet = rechit.energy()/cosh(rhpos.eta());
+    if(rhet<et_thresh_seedrh) continue;
+    if(abs(rhpos.z())>380) continue;
+    double eta = rhpos.eta();
+    double phi = rhpos.phi();
+    
+    bool isiso = true;
+    double etcluster(0);
+    int nbtowerssameeta(0);
+    double timediscriminatorITBH(0),timediscriminatorOTBH(0);
+    double etstrip_phiseedplus1(0), etstrip_phiseedminus1(0);
+
+    //Building the cluster 
+    edm::RefVector<HBHERecHitCollection> bhrhcandidates;
+    for(size_t jhit = 0; jhit<hbherechitcoll->size(); ++ jhit){
+      const HBHERecHit & rechitj = (*hbherechitcoll)[ jhit ];
+      HBHERecHitRef rhRef(hbherechitcoll,jhit);
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
+      if(rhetj<2) continue;
+      if(abs(rhposj.z())>380) continue;
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      if(abs(eta-etaj)>0.4) continue;//This means +/-4 towers in eta 
+      if(abs(deltaPhi(phi,phij))>0.2) continue;//This means +/-2 towers in phi 
+      if(abs(deltaPhi(phi,phij))>0.1&&abs(eta-etaj)<0.2){isiso=false;break;}//The strip should be isolated
+      if(abs(deltaPhi(phi,phij))>0.1)continue;
+      if(abs(deltaPhi(phi,phij))<0.05) nbtowerssameeta++;
+      if(deltaPhi(phi,phij)>0.05) etstrip_phiseedplus1+=rhetj;
+      if(deltaPhi(phi,phij)<-0.05) etstrip_phiseedminus1+=rhetj;
+      
+      etcluster+=rhetj;
+      //Timing discriminator
+      //We assign a weight to the rechit defined as: 
+      //Log10(Et)*f(T,R,Z)
+      //where f(T,R,Z) is the separation curve between halo-like and IP-like times.
+      //The time difference between a deposit from a outgoing IT halo and a deposit coming from a particle emitted at the IP is given by:  
+      //dt= ( - sqrt(R^2+z^2) + |z| )/c 
+      // For OT beam halo, the time difference is: 
+      //dt= ( 25 + sqrt(R^2+z^2) + |z| )/c 
+      //only consider the central part of HB as things get hard at large z.
+      //The best fitted value for R leads to 240 cm (IT) and 330 cm (OT)
+      double rhtj = rechitj.time();
+      timediscriminatorITBH+= TMath::Log10( rhetj )* ( rhtj +0.5*(sqrt(240.*240.+rhposj.z()*rhposj.z()) -abs(rhposj.z()))/30);
+      if(abs(rhposj.z())<300) timediscriminatorOTBH+= TMath::Log10( rhetj )* ( rhtj -0.5*(25-(sqrt(330.*330.+rhposj.z()*rhposj.z()) +abs(rhposj.z()))/30) );
```

Comment:
HCAL reco has a large fraction of ~0 energy hits.
I can't recall if exact zeros still show up.
Please add a check for the hits to be above some small threshold (unless I missed it in upstream selections)